srtcore/smoother.cpp

```diff
@@ -0,0 +1,606 @@
+            LOGC(mglog.Debug) << "FileSmoother: LOSS:PERIOD lseq=" << lossbegin
+                << ", dseq=" << m_iLastDecSeq
+                << ", seqdiff=" << seqdiff
+                << ", deccnt=" << m_iDecCount
+                << ", decrnd=" << m_iDecRandom
+                << ", sndperiod=" << m_dPktSndPeriod << "us";
+        }
+        else
+        {
+            LOGC(mglog.Debug) << "FileSmoother: LOSS:STILL lseq=" << lossbegin
+                << ", dseq=" << m_iLastDecSeq
+                << ", seqdiff=" << seqdiff
+                << ", deccnt=" << m_iDecCount
+                << ", decrnd=" << m_iDecRandom
+                << ", sndperiod=" << m_dPktSndPeriod << "us";
+        }
+
+    }
+
+    void speedupToWindowSize(ETransmissionEvent, EventVariant arg)
+    {
+        ECheckTimerStage stg = arg.get<EventVariant::STAGE>();
+
+        // TEV_INIT is in the beginning of checkTimers(), used
+        // only to synchronize back the values (which is done in updateCC
+        // after emitting the signal).
+        if (stg == TEV_CHT_INIT)
+            return;
+
+        if (m_bSlowStart)
+        {
+            m_bSlowStart = false;
+            if (m_parent->deliveryRate() > 0)
+            {
+                m_dPktSndPeriod = 1000000.0 / m_parent->deliveryRate();
+                LOGC(mglog.Debug) << "FileSmoother: CHKTIMER, SLOWSTART:OFF, sndperiod=" << m_dPktSndPeriod << "us AS mega/rate (rate="
+                    << m_parent->deliveryRate() << ")";
+            }
+            else
+            {
+                m_dPktSndPeriod = m_dCWndSize / (m_parent->RTT() + m_iRCInterval);
+                LOGC(mglog.Debug) << "FileSmoother: CHKTIMER, SLOWSTART:OFF, sndperiod=" << m_dPktSndPeriod << "us AS wndsize/(RTT+RCIV) (wndsize="
+                    << setprecision(6) << m_dCWndSize << " RTT=" << m_parent->RTT() << " RCIV=" << m_iRCInterval << ")";
+            }
+        }
+        else
+        {
+            // XXX This code is a copy of legacy CUDTCC::onTimeout() body.
+            // This part was commented out there already.
+            /*
+               m_dLastDecPeriod = m_dPktSndPeriod;
+               m_dPktSndPeriod = ceil(m_dPktSndPeriod * 2);
+               m_iLastDecSeq = m_iLastAck;
+             */
+        }
+    }
+
+    Smoother::RexmitMethod rexmitMethod() ATR_OVERRIDE
+    {
+        return Smoother::SRM_LATEREXMIT;
+    }
+};
+
+
+#undef SSLOT
+
+template <class Target>
+struct Creator
+{
+    static SmootherBase* Create(CUDT* parent) { return new Target(parent); }
+};
+
+typedef map<string, smoother_create_t*> smoother_base_t;
+smoother_base_t Smoother::registerred_smoothers;
+
+class SmootherBaseInitializer
+{
+public:
+    SmootherBaseInitializer()
+    {
+        Smoother::registerred_smoothers["live"] = Creator<LiveSmoother>::Create;
+        Smoother::registerred_smoothers["file"] = Creator<FileSmoother>::Create;
+    }
+
+} g_smoother_base_initializer;
```

Comment:
Again, I am very much against global non-POD objects being constructed this way unless there is no way around it. These are notoriously hard to debug and since we are just a library this is just not being fair to the app developers. On the other hand, have perfectly good srt_startup(). Can we please move this initializer there? I don't see any loss in functionality from doing so.

Reply:
No, `srt_startup` is the worst possible place to do that. Whether this is initialized in a function called in the beginning, or with global initializer, it makes completely no difference, whereas opening the access to these global objects to all other parts, making them dependent, causes a mess in the design. This object is intentionally not exposed to any other parts and managed exclusively internally. There's also no need for having a singleton because this object is independent on any other parts.

Surely the `srt_startup` (`CUDTUnited::startup` in particular) should be revised because it's doing more than it should. It has its own merit: calling WSAStartup (on Windows) and creating the GC thread - these two things must be done there because they must be done after all initializations of global objects is already finished.

But, on the other hand, you never know when `srt_startup` will be called - the only thing you are certain about is that it will be called before the first use of any SRT sockets (it can be called by user or it's called automatically with the first created socket). You never know the place where it's going to be called, whereas global objects you need initialized already at the very first instruction executed before any user code. Unless you have some idea how to enforce a user to call `srt_startup` as the very first instruction in main, this can't be done this way. All independent global objects must be initialized **before any user code is executed**.

Reply:
I still can't see a problem to solve here, and your solution proposal decreases the design clarity and makes the solution more risky.

Reply:
I disagree. Having an Init() function for an object is much more clear than having a library object auto-initialize with CRT even before main() of the app that is using the lib is called.

Reply:
This `Init()` function still can't be guaranteed to be called before any user code, whereas the runtime initialization of global objects gives that guarantee. The biggest problem with a globally initializing function is that it must work with the prediction that some resources might be already initialized and some others are not, and it must have a method to distinguish them. The global initializer is guaranteed to be called exactly once and the object at the entrance is in uninitialized state. There's also a guarantee that all other resources (usually system ones), which might potentially influence the execution of the initialization code, are in the default state (unless some of them are modified by a runtime initializer, of course, but in SRT nothing like that is being done, and the awareness of this is generally adhered to). It means that having an initializing function is much more risky than leaving the initialization to the global constructor.

C libraries that use this initialization method can only kindly ask a user to call the initialization function before any other function from this library. The risk that a user make a mistake here is a part of this language.

Currently `srt_startup()` is optional to call, and the user may do it in order to have predictable moment when any potentially time-consuming initialization takes place. This change would have to change this function to be now called obligatorily, moreover, in no other place than in the very beginning of `main()`, especially before any function that might mess with any system resources.

Alternatively we can create a separate initialization function, but this is actually even worse.

Reply:
It’s very simple; the user must call Init() before using the library. This is not only common practice, it’s best practice.

Reply:
This can be done this way when there's a new library with no users code yet. That's not the case of SRT. Moreover, this has never been true in C++ - there are lots of runtime-initialized global objects even in the C++ standard library.

Reply:
I think requiring srt_startup() call before any other SRT APIs are called upon is a reasonable requirement. Probably does not matter where in the main() the app calls it but stating that SRT lib is dead until a successful call to srt_startup() sounds very reasonable and should be easy to enforce. I think we are getting somewhere with this, let's discuss this in PM?

Reply:
The so far requirement for `srt_startup()` is that it's optional and may be called when a user wants a predictable initialization moment, so that the user's application doesn't suffer extra delay at the first call to `srt_socket()`. The user is free to call it at any time, including not to call it at all.

To make `srt_startup()` now responsible to initialize global object, we have to change that requirement to that this function must be now called in every application, and not at any time, but strictly in the beginning of `main()`. It is very risky to change this requirement now that there exist code that not necessarily does it this way that we would need in this case (even stating that it's called using the legacy UDT API).

If you really want to have these global objects initialized by a function, the only safe method to do it is to introduce a new function that does the global object initialization - and this function still wouldn't call `srt_startup()` from itself. This is also because the `srt_startup()` initializes the "late resources" (such as extra threads), which constitute a burden for an application, not necessarily needed when occasionally the application doesn't want to use SRT at the moment. Currently a user is free to call `srt_startup()` upon necessity, so if an application doesn't create any SRT sockets at particular scenario, there's no need for this function to be called at all. And of course, currently the `srt_startup()` also requires some objects to be already initialized when it's started, including runtime-initialized.

My point is that this situation has been always the case in SRT (and in UDT), I didn't add any new risk to the application, just added a new brick to the global-runtime-initialization wall. It makes a very little sense to shift the initialization of **just one** object to an initialization function - it would have to be done for **all** global objects used in SRT so far, and this is quite a big refactoring work. And definitely such work isn't a part of this task.

Reply:
Not sure I fully understand the logic with "late resources". As far as I know the library is dead and useless until after srt_startup() is called (directly or indirectly), whether or not some globals have been pre-initialized.

I am aware of s_UDTUnited global that has been inherited from legacy code but that in itself should be straight-forward enough to address should we choose to. Let me do quick accounting of the other bricks in the "global-runtime-initialization wall" and see where they came from and why. We certainly don't want to keep building up the wall if we ever want to see it disappear. Thanks!

Reply:
So, no, this isn't the right understanding. There are still some library facilities that don't use sockets and therefore don't need `srt_startup()` to be called before - but they do need all global objects to be initialized at that time.

Reply:
Resources initialized in this code are twofold: the lightweight and the heavy one. The initialization of the lightweight can happen at any time and no one would ever notice - although at least most of them need to be initialized before any user code is run. The heavy ones may occupy some significant system resource usage at the time of initialization, which means that it matters at what time it is being done. In order not to put a burden to a whole application without the need ("don't pay for what you don't use"), this is usually initialized lazily. The `srt_startup()` initializes the heavy resources, and can be used to shift the initialization moment from the latest possible place where it's needed to some earlier point of execution, if a user can see some advantage in it.